// @parcel/core/lib/Validation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _utils = require("@parcel/utils");

var _logger = _interopRequireWildcard(require("@parcel/logger"));

var _diagnostic = _interopRequireWildcard(require("@parcel/diagnostic"));

var _ParcelConfig = _interopRequireDefault(require("./ParcelConfig"));

var _ConfigLoader = _interopRequireDefault(require("./ConfigLoader"));

var _UncommittedAsset = _interopRequireDefault(require("./UncommittedAsset"));

var _assetUtils = require("./assetUtils");

var _Asset = require("./public/Asset");

var _PluginOptions = _interopRequireDefault(require("./public/PluginOptions"));

var _summarizeRequest = _interopRequireDefault(require("./summarizeRequest"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Validation {
  constructor({
    config,
    dedicatedThread,
    options,
    requests,
    report,
    workerApi
  }) {
    _defineProperty(this, "allAssets", {});

    _defineProperty(this, "allValidators", {});

    _defineProperty(this, "dedicatedThread", void 0);

    _defineProperty(this, "configRequests", void 0);

    _defineProperty(this, "configLoader", void 0);

    _defineProperty(this, "impactfulOptions", void 0);

    _defineProperty(this, "options", void 0);

    _defineProperty(this, "parcelConfig", void 0);

    _defineProperty(this, "report", void 0);

    _defineProperty(this, "requests", void 0);

    _defineProperty(this, "workerApi", void 0);

    this.configLoader = new _ConfigLoader.default({
      options,
      config
    });
    this.dedicatedThread = dedicatedThread !== null && dedicatedThread !== void 0 ? dedicatedThread : false;
    this.options = options;
    this.parcelConfig = config;
    this.report = report;
    this.requests = requests;
    this.workerApi = workerApi;
  }

  async run() {
    let pluginOptions = new _PluginOptions.default(this.options);
    await this.buildAssetsAndValidators();
    await Promise.all(Object.keys(this.allValidators).map(async validatorName => {
      let assets = this.allAssets[validatorName];

      if (assets) {
        let plugin = this.allValidators[validatorName];
        let validatorLogger = new _logger.PluginLogger({
          origin: validatorName
        });
        let validatorResults = [];

        try {
          // If the plugin supports the single-threading validateAll method, pass all assets to it.
          if (plugin.validateAll && this.dedicatedThread) {
            validatorResults = await plugin.validateAll({
              assets: assets.map(asset => new _Asset.Asset(asset)),
              options: pluginOptions,
              logger: validatorLogger,
              resolveConfigWithPath: (configNames, assetFilePath) => (0, _utils.resolveConfig)(this.options.inputFS, assetFilePath, configNames)
            });
          } // Otherwise, pass the assets one-at-a-time
          else if (plugin.validate && !this.dedicatedThread) {
              await Promise.all(assets.map(async asset => {
                let config = null;

                if (plugin.getConfig) {
                  config = await plugin.getConfig({
                    asset: new _Asset.Asset(asset),
                    options: pluginOptions,
                    logger: validatorLogger,
                    resolveConfig: configNames => (0, _utils.resolveConfig)(this.options.inputFS, asset.value.filePath, configNames)
                  });
                }

                let validatorResult = await plugin.validate({
                  asset: new _Asset.Asset(asset),
                  options: pluginOptions,
                  config,
                  logger: validatorLogger
                });
                validatorResults.push(validatorResult);
              }));
            }

          this.handleResults(validatorResults);
        } catch (e) {
          throw new _diagnostic.default({
            diagnostic: (0, _diagnostic.errorToDiagnostic)(e, validatorName)
          });
        }
      }
    }));
  }

  async buildAssetsAndValidators() {
    // Figure out what validators need to be run, and group the assets by the relevant validators.
    await Promise.all(this.requests.map(async request => {
      this.report({
        type: 'validation',
        filePath: request.filePath
      });
      let asset = await this.loadAsset(request);
      let validators = await this.parcelConfig.getValidators(request.filePath);

      for (let validator of validators) {
        this.allValidators[validator.name] = validator.plugin;

        if (this.allAssets[validator.name]) {
          this.allAssets[validator.name].push(asset);
        } else {
          this.allAssets[validator.name] = [asset];
        }
      }
    }));
  }

  handleResults(validatorResults) {
    let warnings = [];
    let errors = [];
    validatorResults.forEach(result => {
      if (result) {
        warnings.push(...result.warnings);
        errors.push(...result.errors);
      }
    });

    if (errors.length > 0) {
      throw new _diagnostic.default({
        diagnostic: errors
      });
    }

    if (warnings.length > 0) {
      _logger.default.warn(warnings);
    }
  }

  async loadAsset(request) {
    let {
      filePath,
      env,
      code,
      sideEffects
    } = request;
    let {
      content,
      size,
      hash,
      isSource
    } = await (0, _summarizeRequest.default)(this.options.inputFS, request); // If the transformer request passed code rather than a filename,
    // use a hash as the base for the id to ensure it is unique.

    let idBase = code != null ? hash : (0, _utils.normalizeSeparators)(_path.default.relative(this.options.projectRoot, filePath));
    return new _UncommittedAsset.default({
      idBase,
      value: (0, _assetUtils.createAsset)({
        idBase,
        filePath: filePath,
        isSource,
        type: _path.default.extname(filePath).slice(1),
        hash,
        env: env,
        stats: {
          time: 0,
          size
        },
        sideEffects: sideEffects
      }),
      options: this.options,
      content
    });
  }

}

exports.default = Validation;
