// @parcel/source-map/dist/node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.init = exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _SourceMap = _interopRequireDefault(require("./SourceMap"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const bindings = require('node-gyp-build')(_path.default.join(__dirname, '..'));

class NodeSourceMap extends _SourceMap.default {
  constructor() {
    super();
    this.sourceMapInstance = new bindings.SourceMap();
  }

  toBuffer() {
    return this.sourceMapInstance.toBuffer();
  }

  findClosestMapping(line, column) {
    let mapping = this.sourceMapInstance.findClosestMapping(line, column);
    let v = this.indexedMappingToStringMapping(mapping);
    return v;
  }

  delete() {}

  static generateEmptyMap(sourceName, sourceContent, lineOffset = 0) {
    let map = new NodeSourceMap();
    map.addEmptyMap(sourceName, sourceContent, lineOffset);
    return map;
  }

}

exports.default = NodeSourceMap;
const init = Promise.resolve();
exports.init = init;
