// @parcel/source-map/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateInlineMap = generateInlineMap;
exports.partialVlqMapToSourceMap = partialVlqMapToSourceMap;

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function generateInlineMap(map) {
  return `data:application/json;charset=utf-8;base64,${Buffer.from(map).toString('base64')}`;
}

function normalisePath(filepath) {
  return filepath.replace(/\\/g, '/');
}

function relatifyPath(filepath, rootDir) {
  // Sourcemaps are made for web, so replace backslashes with regular slashes
  filepath = normalisePath(filepath); // Make root paths relative to the rootDir

  if (filepath[0] === '/') {
    filepath = normalisePath(_path.default.relative(rootDir, filepath));
  } // Prefix relative paths with ./ as it makes it more clear and probably prevents issues


  if (filepath[0] !== '.') {
    filepath = `./${filepath}`;
  }

  return filepath;
}

async function partialVlqMapToSourceMap(map, {
  fs,
  file,
  sourceRoot,
  inlineSources,
  rootDir,
  format = 'string'
}) {
  let root = normalisePath(rootDir || '/');
  let resultMap = { ...map,
    sourcesContent: map.sourcesContent ? map.sourcesContent.map(content => content ? content : null) : [],
    version: 3,
    file,
    sourceRoot
  };
  resultMap.sources = resultMap.sources.map(sourceFilePath => {
    return relatifyPath(sourceFilePath, root);
  });

  if (resultMap.sourcesContent.length < resultMap.sources.length) {
    resultMap.sourcesContent.push(...new Array(resultMap.sources.length - resultMap.sourcesContent.length).fill(null));
  }

  if (fs) {
    resultMap.sourcesContent = await Promise.all(resultMap.sourcesContent.map(async (content, index) => {
      let sourceName = map.sources[index]; // If sourceName starts with `..` it is outside rootDir, in this case we likely cannot access this file from the browser or packaged node_module
      // Because of this we have to include the sourceContent to ensure you can always see the sourcecontent for each mapping.

      if (!content && (inlineSources || sourceName.startsWith('..'))) {
        try {
          return await fs.readFile(_path.default.resolve(root, sourceName), 'utf-8');
        } catch (e) {}
      }

      return content;
    }));
  }

  if (format === 'inline' || format === 'string') {
    let stringifiedMap = JSON.stringify(resultMap);

    if (format === 'inline') {
      return generateInlineMap(stringifiedMap);
    }

    return stringifiedMap;
  }

  return resultMap;
}
