// @parcel/core/lib/public/Target.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.targetToInternalTarget = targetToInternalTarget;
exports.default = void 0;

var _Environment = _interopRequireDefault(require("./Environment"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classPrivateFieldGet(receiver, privateMap) { var descriptor = privateMap.get(receiver); if (!descriptor) { throw new TypeError("attempted to get private field on non-instance"); } if (descriptor.get) { return descriptor.get.call(receiver); } return descriptor.value; }

function _classPrivateFieldSet(receiver, privateMap, value) { var descriptor = privateMap.get(receiver); if (!descriptor) { throw new TypeError("attempted to set private field on non-instance"); } if (descriptor.set) { descriptor.set.call(receiver, value); } else { if (!descriptor.writable) { throw new TypeError("attempted to set read only private field"); } descriptor.value = value; } return value; }

const internalTargetToTarget = new WeakMap();

const _targetToInternalTarget = new WeakMap();

function targetToInternalTarget(target) {
  return (0, _nullthrows.default)(_targetToInternalTarget.get(target));
}

class Target {
  // TargetValue
  constructor(target) {
    _target.set(this, {
      writable: true,
      value: void 0
    });

    let existing = internalTargetToTarget.get(target);

    if (existing != null) {
      return existing;
    }

    _classPrivateFieldSet(this, _target, target);

    _targetToInternalTarget.set(this, target);

    internalTargetToTarget.set(target, this);
  }

  get distEntry() {
    return _classPrivateFieldGet(this, _target).distEntry;
  }

  get distDir() {
    return _classPrivateFieldGet(this, _target).distDir;
  }

  get env() {
    return new _Environment.default(_classPrivateFieldGet(this, _target).env);
  }

  get sourceMap() {
    return _classPrivateFieldGet(this, _target).sourceMap;
  }

  get name() {
    return _classPrivateFieldGet(this, _target).name;
  }

  get publicUrl() {
    return _classPrivateFieldGet(this, _target).publicUrl;
  }

  get loc() {
    return _classPrivateFieldGet(this, _target).loc;
  }

}

exports.default = Target;

var _target = new WeakMap();
