// @parcel/reporter-dev-server/lib/HMRServer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ws = _interopRequireDefault(require("ws"));

var _utils = require("@parcel/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class HMRServer {
  constructor(options) {
    _defineProperty(this, "wss", void 0);

    _defineProperty(this, "unresolvedError", null);

    _defineProperty(this, "options", void 0);

    this.options = options;
  }

  start() {
    let websocketOptions = {
      /*verifyClient: info => {
          if (!this.options.host) return true;
           let originator = new URL(info.origin);
          return this.options.host === originator.hostname;
        }*/
    };

    if (this.options.devServer) {
      websocketOptions.server = this.options.devServer;
    } else if (this.options.port) {
      websocketOptions.port = this.options.port;
    }

    this.wss = new _ws.default.Server(websocketOptions);
    this.wss.on('connection', ws => {
      ws.onerror = this.handleSocketError;

      if (this.unresolvedError) {
        ws.send(JSON.stringify(this.unresolvedError));
      }
    });
    this.wss.on('error', this.handleSocketError);
    return this.wss._server.address().port;
  }

  stop() {
    this.wss.close();
  }

  async emitError(options, diagnostics) {
    let renderedDiagnostics = await Promise.all(diagnostics.map(d => (0, _utils.prettyDiagnostic)(d, options))); // store the most recent error so we can notify new connections
    // and so we can broadcast when the error is resolved

    this.unresolvedError = {
      type: 'error',
      diagnostics: {
        ansi: renderedDiagnostics,
        html: renderedDiagnostics.map(d => {
          return {
            message: (0, _utils.ansiHtml)(d.message),
            stack: (0, _utils.ansiHtml)(d.stack),
            codeframe: (0, _utils.ansiHtml)(d.codeframe),
            hints: d.hints.map(hint => (0, _utils.ansiHtml)(hint))
          };
        })
      }
    };
    this.broadcast(this.unresolvedError);
  }

  async emitUpdate(event) {
    this.unresolvedError = null;
    let changedAssets = Array.from(event.changedAssets.values());
    if (changedAssets.length === 0) return;
    let assets = await Promise.all(changedAssets.map(async asset => {
      let dependencies = event.bundleGraph.getDependencies(asset);
      let depsByBundle = {};

      for (let bundle of event.bundleGraph.findBundlesWithAsset(asset)) {
        let deps = {};

        for (let dep of dependencies) {
          let resolved = event.bundleGraph.getDependencyResolution(dep, bundle);

          if (resolved) {
            deps[dep.moduleSpecifier] = resolved.id;
          }
        }

        depsByBundle[bundle.id] = deps;
      }

      return {
        id: asset.id,
        type: asset.type,
        output: await asset.getCode(),
        envHash: (0, _utils.md5FromObject)(asset.env),
        depsByBundle
      };
    }));
    this.broadcast({
      type: 'update',
      assets: assets
    });
  }

  handleSocketError(err) {
    if (err.code === 'ECONNRESET') {
      // This gets triggered on page refresh, ignore this
      return;
    }

    this.options.logger.warn({
      origin: '@parcel/reporter-dev-server',
      message: `[${err.code}]: ${err.message}`,
      stack: err.stack
    });
  }

  broadcast(msg) {
    const json = JSON.stringify(msg);

    for (let ws of this.wss.clients) {
      ws.send(json);
    }
  }

}

exports.default = HMRServer;
