// @parcel/reporter-dev-server/lib/Server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var _events = _interopRequireDefault(require("events"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _path = _interopRequireDefault(require("path"));

var _url = _interopRequireWildcard(require("url"));

var _utils = require("@parcel/utils");

var _serverErrors = _interopRequireDefault(require("./serverErrors"));

var _fs = _interopRequireDefault(require("fs"));

var _ejs = _interopRequireDefault(require("ejs"));

var _connect = _interopRequireDefault(require("connect"));

var _httpProxyMiddleware = _interopRequireDefault(require("http-proxy-middleware"));

var _mime = _interopRequireDefault(require("mime"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function setHeaders(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, HEAD, PUT, PATCH, POST, DELETE');
  res.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Content-Type');
}

const SOURCES_ENDPOINT = '/__parcel_source_root';

const TEMPLATE_404 = _fs.default.readFileSync(_path.default.join(__dirname, 'templates/404.html'), 'utf8');

const TEMPLATE_500 = _fs.default.readFileSync(_path.default.join(__dirname, 'templates/500.html'), 'utf8');

class Server extends _events.default {
  constructor(options) {
    super();

    _defineProperty(this, "pending", void 0);

    _defineProperty(this, "options", void 0);

    _defineProperty(this, "rootPath", void 0);

    _defineProperty(this, "bundleGraph", void 0);

    _defineProperty(this, "errors", void 0);

    _defineProperty(this, "stopServer", void 0);

    this.options = options;

    try {
      this.rootPath = new _url.URL(options.publicUrl).pathname;
    } catch (e) {
      this.rootPath = options.publicUrl;
    }

    this.pending = true;
    this.bundleGraph = null;
    this.errors = null;
  }

  buildStart() {
    this.pending = true;
  }

  buildSuccess(bundleGraph) {
    this.bundleGraph = bundleGraph;
    this.errors = null;
    this.pending = false;
    this.emit('bundled');
  }

  async buildError(options, diagnostics) {
    this.pending = false;
    this.errors = await Promise.all(diagnostics.map(async d => {
      let ansiDiagnostic = await (0, _utils.prettyDiagnostic)(d, options);
      return {
        message: (0, _utils.ansiHtml)(ansiDiagnostic.message),
        stack: ansiDiagnostic.codeframe ? (0, _utils.ansiHtml)(ansiDiagnostic.codeframe) : (0, _utils.ansiHtml)(ansiDiagnostic.stack),
        hints: ansiDiagnostic.hints.map(hint => (0, _utils.ansiHtml)(hint))
      };
    }));
  }

  respond(req, res) {
    let {
      pathname
    } = _url.default.parse(req.originalUrl || req.url);

    if (pathname == null) {
      pathname = '/';
    }

    if (this.errors) {
      return this.send500(req, res);
    } else if (_path.default.extname(pathname) === '') {
      // If the URL doesn't start with the public path, or the URL doesn't
      // have a file extension, send the main HTML bundle.
      return this.sendIndex(req, res);
    } else if (pathname.startsWith(SOURCES_ENDPOINT)) {
      req.url = pathname.slice(SOURCES_ENDPOINT.length);
      return this.serve(this.options.inputFS, this.options.projectRoot, req, res, () => this.send404(req, res));
    } else if (pathname.startsWith(this.rootPath)) {
      // Otherwise, serve the file from the dist folder
      req.url = this.rootPath === '/' ? pathname : pathname.slice(this.rootPath.length);
      return this.serveDist(req, res, () => this.sendIndex(req, res));
    } else {
      return this.send404(req, res);
    }
  }

  sendIndex(req, res) {
    if (this.bundleGraph) {
      // If the main asset is an HTML file, serve it
      let htmlBundle = this.bundleGraph.traverseBundles((bundle, context, {
        stop
      }) => {
        if (bundle.type !== 'html' || !bundle.isEntry) return;

        if (!context) {
          context = bundle;
        }

        if (context && bundle.filePath && bundle.filePath.endsWith('index.html')) {
          stop();
          return bundle;
        }
      });

      if (htmlBundle) {
        req.url = `/${_path.default.relative(this.options.distDir, (0, _nullthrows.default)(htmlBundle.filePath))}`;
        this.serveDist(req, res, () => this.send404(req, res));
      } else {
        this.send404(req, res);
      }
    } else {
      this.send404(req, res);
    }
  }

  serveDist(req, res, next) {
    return this.serve(this.options.outputFS, this.options.distDir, req, res, next);
  }

  async serve(fs, root, req, res, next) {
    if (req.method !== 'GET' && req.method !== 'HEAD') {
      // method not allowed
      res.statusCode = 405;
      res.setHeader('Allow', 'GET, HEAD');
      res.setHeader('Content-Length', '0');
      res.end();
      return;
    }

    try {
      var filePath = _url.default.parse(req.url).pathname || '';
      filePath = decodeURIComponent(filePath);
    } catch (err) {
      return this.sendError(res, 400);
    }

    filePath = _path.default.normalize('.' + _path.default.sep + filePath); // malicious path

    if (filePath.includes(_path.default.sep + '..' + _path.default.sep)) {
      return this.sendError(res, 403);
    } // join / normalize from the root dir


    if (!_path.default.isAbsolute(filePath)) {
      filePath = _path.default.normalize(_path.default.join(root, filePath));
    }

    try {
      var stat = await fs.stat(filePath);
    } catch (err) {
      if (err.code === 'ENOENT') {
        return next(req, res);
      }

      return this.sendError(res, 500);
    } // Fall back to next handler if not a file


    if (!stat || !stat.isFile()) {
      return next(req, res);
    }

    setHeaders(res);
    res.setHeader('Content-Length', '' + stat.size);

    let mimeType = _mime.default.getType(filePath);

    if (mimeType != null) {
      res.setHeader('Content-Type', mimeType + '; charset=utf-8');
    }

    if (req.method === 'HEAD') {
      res.end();
      return;
    }

    return new Promise((resolve, reject) => {
      fs.createReadStream(filePath).pipe(res).on('finish', resolve).on('error', reject);
    });
  }

  sendError(res, statusCode) {
    res.statusCode = statusCode;
    setHeaders(res);
    res.end();
  }

  send404(req, res) {
    res.statusCode = 404;
    setHeaders(res);
    res.end(TEMPLATE_404);
  }

  send500(req, res) {
    setHeaders(res);
    res.setHeader('Content-Type', 'text/html; charset=utf-8');
    res.writeHead(500);

    if (this.errors) {
      return res.end(_ejs.default.render(TEMPLATE_500, {
        errors: this.errors
      }));
    }
  }

  logAccessIfVerbose(req) {
    this.options.logger.verbose({
      message: `Request: ${req.headers.host}${req.originalUrl || req.url}`
    });
  }
  /**
   * Load proxy table from package.json and apply them.
   */


  async applyProxyTable(app) {
    // avoid skipping project root
    const fileInRoot = _path.default.join(this.options.projectRoot, '_');

    const pkg = await (0, _utils.loadConfig)(this.options.inputFS, fileInRoot, ['.proxyrc.js', '.proxyrc']);

    if (!pkg || !pkg.config || !pkg.files) {
      return this;
    }

    const cfg = pkg.config;

    const filename = _path.default.basename(pkg.files[0].filePath);

    if (filename === '.proxyrc.js') {
      if (typeof cfg !== 'function') {
        this.options.logger.warn({
          message: "Proxy configuration file '.proxyrc.js' should export a function. Skipping..."
        });
        return this;
      }

      cfg(app);
    } else if (filename === '.proxyrc') {
      if (typeof cfg !== 'object') {
        this.options.logger.warn({
          message: "Proxy table in '.proxyrc' should be of object type. Skipping..."
        });
        return this;
      }

      for (const [context, options] of Object.entries(cfg)) {
        // each key is interpreted as context, and value as middleware options
        app.use((0, _httpProxyMiddleware.default)(context, options));
      }
    }

    return this;
  }

  async start() {
    const app = (0, _connect.default)();
    await this.applyProxyTable(app);
    app.use((req, res) => {
      this.logAccessIfVerbose(req);

      const response = () => this.respond(req, res); // Wait for the parcelInstance to finish bundling if needed


      if (this.pending) {
        this.once('bundled', response);
      } else {
        response();
      }
    });
    let {
      server,
      stop
    } = await (0, _utils.createHTTPServer)({
      cacheDir: this.options.cacheDir,
      https: this.options.https,
      inputFS: this.options.inputFS,
      listener: app,
      outputFS: this.options.outputFS,
      host: this.options.host
    });
    this.stopServer = stop;
    server.listen(this.options.port, this.options.host);
    return new Promise((resolve, reject) => {
      server.once('error', err => {
        this.options.logger.error({
          message: (0, _serverErrors.default)(err, this.options.port)
        });
        reject(err);
      });
      server.once('listening', () => {
        resolve(server);
      });
    });
  }

  async stop() {
    (0, _assert.default)(this.stopServer != null);
    await this.stopServer();
    this.stopServer = null;
  }

}

exports.default = Server;
